include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -637,6 +655,9 @@ namespace VectorTools
         if (cell2->active() && !cell2->is_locally_owned())
           continue;
 
+        Assert(internal::is_locally_owned(cell1) && internal::is_locally_owned(cell2),
+               ExcMessage("The two Triangulations are required to have the same parallel partitioning."));
+
```

Comment:
I think this is too late. You've already skipped the interesting cells for which this may not be true above.